@alignCast() causes assertion failure when casting from type with no in-memory bits

Compiler output:
```
Assertion failed at /deps/zig/src/analyze.cpp:615 in get_pointer_to_type_extra2. This is a bug in the Zig compiler.
Unable to dump stack trace: debug info stripped
Aborted (core dumped)
```

The compiler crashes on the following test. It should fail.
```
pub const RawCookie = *struct{};

pub fn cookieAs(self: RawCookie, comptime T: type) *T {
    return @ptrCast(*T, @alignCast(@alignOf(*T), self));
}

test "Bug" {
    const ThingType = struct {
        name: []const u8 = "<null>",
        n_links: u64 = 0,
    };

    var thing = ThingType{};

    var cookie: ?RawCookie = @ptrCast(RawCookie, &thing);
    var unwrapped_cookie = cookieAs(cookie.?, ThingType);
}
```

Expected output should be similar to:
```
./ok2.zig:4:12: error: '*RawCookie' and '*ThingType' do not have the same in-memory representation
    return @ptrCast(*T, self);
           ^
./ok2.zig:16:36: note: called from here
    var unwrapped_cookie = cookieAs(cookie.?, ThingType);
                                   ^
./ok2.zig:7:12: note: called from here
test "Bug" {
           ^
./ok2.zig:4:25: note: '*RawCookie' has no in-memory bits
    return @ptrCast(*T, self);
                        ^
./ok2.zig:4:21: note: '*ThingType' has in-memory bits
    return @ptrCast(*T, self);
                    ^
```

The tests used are attached: [tests.zip](https://github.com/ziglang/zig/files/5009741/tests.zip).

- bug.zig, bug2.zig: both crash the compiler
- ok.zig: passes (`RawCookie` is a type that has in-memory bits)
- ok2.zig: fails (`@alignCast` not used)

--- a/src/ir.cpp
+++ b/src/ir.cpp
@@ -29098,6 +29098,19 @@ static IrInstGen *ir_align_cast(IrAnalyze *ira, IrInstGen *target, uint32_t alig
     ZigType *result_type;
     uint32_t old_align_bytes;
 
+    ZigType *actual_ptr = target_type;
+    if (actual_ptr->id == ZigTypeIdOptional) {
+        actual_ptr = actual_ptr->data.maybe.child_type;
+    } else if (is_slice(actual_ptr)) {
+        actual_ptr = actual_ptr->data.structure.fields[slice_ptr_index]->type_entry;
+    }
+
+    if (safety_check_on && !type_has_bits(ira->codegen, actual_ptr)) {
+        ir_add_error(ira, &target->base,
+            buf_sprintf("cannot adjust alignment of zero sized type '%s'", buf_ptr(&target_type->name)));
+        return ira->codegen->invalid_inst_gen;
+    }
+
     if (target_type->id == ZigTypeIdPointer) {
         result_type = adjust_ptr_align(ira->codegen, target_type, align_bytes);
         if ((err = resolve_ptr_align(ira, target_type, &old_align_bytes)))

--- a/test/compile_errors.zig
+++ b/test/compile_errors.zig
@@ -2,6 +2,32 @@ const tests = @import("tests.zig");
 const std = @import("std");
 
 pub fn addCases(cases: *tests.CompileErrorContext) void {
+    cases.addTest("@alignCast of zero sized types",
+        \\export fn foo() void {
+        \\    const a: *void = undefined;
+        \\    _ = @alignCast(2, a);
+        \\}
+        \\export fn bar() void {
+        \\    const a: ?*void = undefined;
+        \\    _ = @alignCast(2, a);
+        \\}
+        \\export fn baz() void {
+        \\    const a: []void = undefined;
+        \\    _ = @alignCast(2, a);
+        \\}
+        \\export fn qux() void {
+        \\    const a = struct {
+        \\        fn a(comptime b: u32) void {}
+        \\    }.a;
+        \\    _ = @alignCast(2, a);
+        \\}
+    , &[_][]const u8{
+        "tmp.zig:3:23: error: cannot adjust alignment of zero sized type '*void'",
+        "tmp.zig:7:23: error: cannot adjust alignment of zero sized type '?*void'",
+        "tmp.zig:11:23: error: cannot adjust alignment of zero sized type '[]void'",
+        "tmp.zig:17:23: error: cannot adjust alignment of zero sized type 'fn(u32) anytype'",
+    });
+
     cases.addTest("invalid pointer with @Type",
         \\export fn entry() void {
         \\    _ = @Type(.{ .Pointer = .{

